File: src/ruby/basic/sudoku.rb
require 'socket'

class Grid
    def initialize(cells)
        @cells = cells
    end
    
    def set_value(row, col, value)
        @cells[row][col] = value
    end

    def is_valid(row_index, col_index, value)
        for row_loop_i in 0..8
            if @cells[row_loop_i][col_index] == value then
                return false
            end
        end

        for col_loop_i in 0..8
            if @cells[row_index][col_loop_i] == value then
                return false
            end
        end
        
        row_start = (row_index / 3).floor() * 3
        col_start = (col_index / 3).floor() * 3

        for row_loop_i in 0..2
            for col_loop_i in 0..2
                if @cells[row_start + row_loop_i][col_start + col_loop_i] == value then
                    return false
                end
            end
        end

        return true
    end

    def raw_output
      @cells.map(&:join).join
    end
    
    def output()
        @cells
    end

    def solve() 
        @cells.each_with_index do |row, rowIndex| 
            @cells[rowIndex].each_with_index do |col, colIndex|
                if @cells[rowIndex][colIndex] == 0 then
                    for value in 1..9 
                        if self.is_valid(rowIndex, colIndex, value) then
                            self.set_value(rowIndex, colIndex, value)

                            if self.solve() then
                                return true
                            end
                        end
                    end

                    self.set_value(rowIndex, colIndex, 0)
                    return false
                end
            end
        end

        return true
    end

end


def solve(string)
    array = string.chomp.scan(/.{9}/).map do |row|
      row.split("").map(&:to_i)
    end

    grid = Grid.new(array)
    grid.solve
    grid.raw_output
end

if ENV['SUDOKU']
    print solve(ENV['SUDOKU']), "\n"
    exit(0)
end

s = UNIXSocket.new("/tmp/sudoku.sock")
s.send "ruby:basic", 0

loop do
    data = s.read(81)
    exit if data.nil? || data.empty?
    s.send solve(data), 0
end

File: src/php/basic/sudoku.php
<?php

class Grid
{
    protected $data = [];
    
    public function __construct(array $data)
    {
        $this->data = $data;
    }

    public function isValid(int $row, int $col, int $value) : bool
    {
        // Rows first!
        for ($colI = 0; $colI < 9; $colI++) {
            if ($this->data[$row][$colI] === $value) {
                return false;
            }
        }

        // Then columns
        for ($rowI = 0; $rowI < 9; $rowI++) {
            if ($this->data[$rowI][$col] === $value) {
                return false;
            }
        }

        // Now, work out boxes
        $rowStart = (int)floor($row / 3) * 3;
        $colStart = (int)floor($col / 3) * 3;
        for ($rowI = $rowStart; $rowI < $rowStart + 3; $rowI++) {
            for ($colI = $colStart; $colI < $colStart + 3; $colI++) {
                if ($this->data[$rowI][$colI] == $value) {
                    return false;
                }
            }
        }

        return true;
    }

    public function set(int $row, int $col, ?int $value)
    {
        $this->data[$row][$col] = $value;
    }

    public function solve() : bool
    {
        for ($row = 0; $row < 9; $row++) {
            for ($col = 0; $col < 9; $col++) {
                if (!isset($this->data[$row][$col])) {
                    var_dump($this->data);
                    die();
                }
                if ($this->data[$row][$col] !== 0) {
                    continue;
                }

                for ($num = 1; $num <= 9; $num ++){
                    if ($this->isValid($row, $col, $num)) {
                        $this->set($row, $col, $num);
                        if ($this->solve()) {
                            return true;
                        }
                    }
                }

                $this->set($row, $col, 0);
                return false;
            }
        }
        return true;
    }

    public function output()
    {
        return implode("", array_map(fn($v) => implode("", $v), $this->data));
    }
}

function solve(string $string) {
    $data = array_chunk(array_map(fn($v) => (int)$v, str_split($string)), 9);
    $grid = new Grid($data);
    $grid->solve();
    return $grid->output();
}

if ($_ENV["SUDOKU"]) {
    $content = trim($_ENV["SUDOKU"]);
    echo solve($content) . "\n";
    exit(0);
}

$sock = stream_socket_client('unix:////tmp/sudoku.sock', $errno, $errstr);

fwrite($sock, 'php:basic');
while ($sock) {
    $content = fread($sock, 81);
    if ($content === "") {
        break;
    }

    fwrite($sock, solve($content));
}
fclose($sock);
